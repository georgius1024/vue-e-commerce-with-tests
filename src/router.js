import * as VueRouter from "vue-router";

import Home from "@/views/Home.vue";
import Category from "@/views/Category.vue";
import Product from "@/views/Product.vue";
import Checkout from "@/views/Checkout.vue";
import Error from "@/views/Error404.vue";
const routes = [
  { path: "/", name: "home", component: Home },
  { path: "/:category", name: "category", component: Category, props: true },
  { path: "/:category/:id", name: "product", component: Product, props: true },
  { path: "/checkout", name: "checkout", component: Checkout },
  { path: "/:pathMatch(.*)*", component: Error },
];
export default VueRouter.createRouter({
  history: VueRouter.createWebHashHistory(),
  routes,
});
