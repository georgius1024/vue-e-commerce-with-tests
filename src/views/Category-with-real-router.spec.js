import { describe, it, expect, vi, beforeEach } from "vitest";
import { mount, flushPromises } from "@vue/test-utils";
import component from "@/views/Category.vue";
import router from "@/router";

vi.mock("@/services/products", () => ({
  default: vi.fn().mockResolvedValue({
    products: global.productsMock,
  }),
}));

describe("Category page with real router", () => {
  beforeEach(async () => {
    router.push("/");
    await router.isReady();
  });
  const buildWrapper = () => {
    const propsData = { category: "General-Products" };

    return mount(component, {
      propsData,
      global: {
        plugins: [router],
      },
    });
  };
  it("mounts without error", async () => {
    const wrapper = buildWrapper();
    expect(wrapper.exists).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
    await flushPromises();
    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.html()).toContain("100");
    expect(wrapper.html()).toContain("product-3");
  });

  it("click on product preview navigates to product", async () => {
    const push = vi.spyOn(router, "push");
    const wrapper = buildWrapper();
    expect(wrapper.exists).toBeTruthy();
    await flushPromises();
    await wrapper.find('[data-testid="product-preview-101"').trigger("click");
    expect(push).toBeCalled();
  });
});
