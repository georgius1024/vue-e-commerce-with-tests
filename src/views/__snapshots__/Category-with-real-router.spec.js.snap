// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Category page with real router > mounts without error 1`] = `
"<div class=\\"container\\" category=\\"General-Products\\">
  <ol class=\\"breadcrumb\\">
    <li class=\\"breadcrumb-item\\"><a aria-current=\\"page\\" href=\\"#/\\" class=\\"router-link-active router-link-exact-active\\">Categories</a></li>
    <li class=\\"breadcrumb-item active text-capitalize\\"></li>
  </ol>
  <h1>Products in category </h1>
  <div class=\\"d-flex flex-row flex-wrap\\"></div>
</div>"
`;

exports[`Category page with real router > mounts without error 2`] = `
"<div class=\\"container\\" category=\\"General-Products\\">
  <ol class=\\"breadcrumb\\">
    <li class=\\"breadcrumb-item\\"><a aria-current=\\"page\\" href=\\"#/\\" class=\\"router-link-active router-link-exact-active\\">Categories</a></li>
    <li class=\\"breadcrumb-item active text-capitalize\\"></li>
  </ol>
  <h1>Products in category </h1>
  <div class=\\"d-flex flex-row flex-wrap\\">
    <div data-v-84d1069b=\\"\\" class=\\"card pointer hover-up mb-2 me-2\\" data-testid=\\"product-preview-101\\"><img data-v-84d1069b=\\"\\" class=\\"card-img-top\\" alt=\\"preview\\">
      <div data-v-84d1069b=\\"\\" class=\\"card-body\\">
        <h5 data-v-84d1069b=\\"\\" class=\\"card-title\\">product-1</h5>
        <p data-v-84d1069b=\\"\\" class=\\"card-text\\">Lorem ipsum 1</p><span data-v-84d1069b=\\"\\" class=\\"float-end badge rounded-pill fs-5 bg-primary mt-3 me-3\\"> $100</span>
      </div>
    </div>
    <div data-v-84d1069b=\\"\\" class=\\"card pointer hover-up mb-2 me-2\\" data-testid=\\"product-preview-102\\"><img data-v-84d1069b=\\"\\" class=\\"card-img-top\\" alt=\\"preview\\">
      <div data-v-84d1069b=\\"\\" class=\\"card-body\\">
        <h5 data-v-84d1069b=\\"\\" class=\\"card-title\\">product-2</h5>
        <p data-v-84d1069b=\\"\\" class=\\"card-text\\">Lorem ipsum 2</p><span data-v-84d1069b=\\"\\" class=\\"float-end badge rounded-pill fs-5 bg-primary mt-3 me-3\\"> $200</span>
      </div>
    </div>
    <div data-v-84d1069b=\\"\\" class=\\"card pointer hover-up mb-2 me-2\\" data-testid=\\"product-preview-103\\"><img data-v-84d1069b=\\"\\" class=\\"card-img-top\\" alt=\\"preview\\">
      <div data-v-84d1069b=\\"\\" class=\\"card-body\\">
        <h5 data-v-84d1069b=\\"\\" class=\\"card-title\\">product-3</h5>
        <p data-v-84d1069b=\\"\\" class=\\"card-text\\">Lorem ipsum 3</p><span data-v-84d1069b=\\"\\" class=\\"float-end badge rounded-pill fs-5 bg-primary mt-3 me-3\\"> $300</span>
      </div>
    </div>
  </div>
</div>"
`;
